File: android/app/src/main/java/com/nxtend/team35/yubiboard/ui/ProductionModels.kt
package com.nxtend.team35.yubiboard.ui

import com.nxtend.team35.yubiboard.network.ConnectionSnapshot
import com.nxtend.team35.yubiboard.network.ConnectionStatus
import com.nxtend.team35.yubiboard.protocol.CaptureMode
import com.nxtend.team35.yubiboard.vision.DetectedMarker
import com.nxtend.team35.yubiboard.vision.LandmarkPoint
import com.nxtend.team35.yubiboard.vision.MarkerDetectionResult

enum class CameraUiState {
    PERMISSION_REQUIRED,
    PERMISSION_DENIED,
    STARTING,
    READY,
    ERROR,
}

sealed interface CalibrationUiState {
    data object Inactive : CalibrationUiState
    data object PlacementWaiting : CalibrationUiState
    data class FindingMarkers(val found: Int) : CalibrationUiState
    data class Stabilizing(val current: Int, val required: Int) : CalibrationUiState
    data object WaitingForPc : CalibrationUiState
    data class RetryRequired(val reason: CalibrationRetryReason) : CalibrationUiState
    data object Complete : CalibrationUiState
}

fun calibrationUiStateAfterFrame(
    current: CalibrationUiState,
    result: MarkerDetectionResult,
): CalibrationUiState = when {
    current in setOf(CalibrationUiState.WaitingForPc, CalibrationUiState.Complete) -> current
    result.stable -> CalibrationUiState.WaitingForPc
    current == CalibrationUiState.PlacementWaiting && result.markers.isEmpty() ->
        CalibrationUiState.PlacementWaiting
    current is CalibrationUiState.RetryRequired && result.markers.isEmpty() -> current
    result.markers.size < 4 -> CalibrationUiState.FindingMarkers(result.markers.size)
    else -> CalibrationUiState.Stabilizing(result.stableFrameCount, result.requiredStableFrames)
}

enum class CalibrationRetryReason {
    MARKERS_NOT_VISIBLE,
    INVALID_GEOMETRY,
    UNSTABLE,
    SCREEN_MISMATCH,
    INTERNAL_ERROR,
    UNKNOWN,
}

enum class TrackingUiState {
    INACTIVE,
    READY_NO_HAND,
    CANDIDATE,
    TRACKING,
    TEMPORARILY_LOST,
    LONG_LOST,
}

data class ProductionUiState(
    val camera: CameraUiState = CameraUiState.STARTING,
    val connection: ConnectionSnapshot = ConnectionSnapshot(ConnectionStatus.DISCONNECTED),
    val captureMode: CaptureMode = CaptureMode.TRACKING,
    val calibration: CalibrationUiState = CalibrationUiState.Inactive,
    val tracking: TrackingUiState = TrackingUiState.INACTIVE,
    val indexTip: LandmarkPoint? = null,
    val markers: List<DetectedMarker> = emptyList(),
    val sourceWidth: Int = 0,
    val sourceHeight: Int = 0,
    val notice: String? = null,
)

enum class ProductionStage {
    CAMERA_PERMISSION,
    CAMERA_ERROR,
    CONNECT,
    CONNECTING,
    AUTO_CONNECTING,
    CONNECTION_ERROR,
    RECONNECTING,
    CALIBRATION,
    READY,
}

/**
 * Visual families used by the production guide panel.
 *
 * Keeping this mapping separate from the composables makes the redesigned UI follow the
 * existing connection/calibration/tracking state machine without introducing a second source
 * of truth for app behaviour.
 */
enum class ProductionVisualState {
    CAMERA_PERMISSION,
    CAMERA_ERROR,
    CONNECTION_FORM,
    CONNECTION_PROGRESS,
    RECONNECTING,
    PLACEMENT,
    CALIBRATION_PROGRESS,
    READY_IDLE,
    READY_ACTIVE,
}

fun ProductionUiState.stage(): ProductionStage = when {
    camera == CameraUiState.PERMISSION_REQUIRED || camera == CameraUiState.PERMISSION_DENIED ->
        ProductionStage.CAMERA_PERMISSION
    camera == CameraUiState.ERROR -> ProductionStage.CAMERA_ERROR
    connection.status == ConnectionStatus.ERROR -> ProductionStage.CONNECTION_ERROR
    connection.status == ConnectionStatus.RECONNECTING -> ProductionStage.RECONNECTING
    connection.status in setOf(ConnectionStatus.CONNECTING, ConnectionStatus.AWAITING_ACK) ->
        if (connection.automatic) ProductionStage.AUTO_CONNECTING else ProductionStage.CONNECTING
    connection.status == ConnectionStatus.DISCONNECTED -> ProductionStage.CONNECT
    captureMode == CaptureMode.CALIBRATION -> ProductionStage.CALIBRATION
    else -> ProductionStage.READY
}

fun ProductionUiState.visualState(): ProductionVisualState = when (stage()) {
    ProductionStage.CAMERA_PERMISSION -> ProductionVisualState.CAMERA_PERMISSION
    ProductionStage.CAMERA_ERROR -> ProductionVisualState.CAMERA_ERROR
    ProductionStage.CONNECT, ProductionStage.CONNECTION_ERROR ->
        ProductionVisualState.CONNECTION_FORM
    ProductionStage.CONNECTING, ProductionStage.AUTO_CONNECTING ->
        ProductionVisualState.CONNECTION_PROGRESS
    ProductionStage.RECONNECTING -> ProductionVisualState.RECONNECTING
    ProductionStage.CALIBRATION -> when (calibration) {
        CalibrationUiState.PlacementWaiting -> ProductionVisualState.PLACEMENT
        else -> ProductionVisualState.CALIBRATION_PROGRESS
    }
    ProductionStage.READY -> when (tracking) {
        TrackingUiState.CANDIDATE,
        TrackingUiState.TRACKING,
        TrackingUiState.TEMPORARILY_LOST,
        TrackingUiState.LONG_LOST -> ProductionVisualState.READY_ACTIVE
        else -> ProductionVisualState.READY_IDLE
    }
}

File: ios/Screact/UI/ProductionModels.swift
import Foundation

/// Port of Android ProductionModels: the UI state machine that drives the guide panel. The stage /
/// visualState mapping is a faithful translation so the SwiftUI screen follows the exact same
/// connection / calibration / tracking transitions as Compose.

enum CameraUiState {
    case permissionRequired
    case permissionDenied
    case starting
    case ready
    case error
}

enum CalibrationRetryReason {
    case markersNotVisible
    case invalidGeometry
    case unstable
    case screenMismatch
    case internalError
    case unknown
}

enum CalibrationUiState: Equatable {
    case inactive
    case placementWaiting
    case findingMarkers(found: Int)
    case stabilizing(current: Int, required: Int)
    case waitingForPc
    case retryRequired(reason: CalibrationRetryReason)
    case complete
}

func calibrationUiStateAfterFrame(_ current: CalibrationUiState,
                                  _ result: MarkerDetectionResult) -> CalibrationUiState {
    if current == .waitingForPc || current == .complete { return current }
    if result.stable { return .waitingForPc }
    if current == .placementWaiting && result.markers.isEmpty { return .placementWaiting }
    if case .retryRequired = current, result.markers.isEmpty { return current }
    if result.markers.count < 4 { return .findingMarkers(found: result.markers.count) }
    return .stabilizing(current: result.stableFrameCount, required: result.requiredStableFrames)
}

enum TrackingUiState {
    case inactive
    case readyNoHand
    case candidate
    case tracking
    case temporarilyLost
    case longLost
}

enum ProductionStage {
    case cameraPermission
    case cameraError
    case connect
    case connecting
    case autoConnecting
    case connectionError
    case reconnecting
    case calibration
    case ready
}

enum ProductionVisualState {
    case cameraPermission
    case cameraError
    case connectionForm
    case connectionProgress
    case reconnecting
    case placement
    case calibrationProgress
    case readyIdle
    case readyActive
}

struct ProductionUiState {
    var camera: CameraUiState = .starting
    var connection: ConnectionSnapshot = ConnectionSnapshot(.disconnected)
    var captureMode: CaptureMode = .tracking
    var calibration: CalibrationUiState = .inactive
    var tracking: TrackingUiState = .inactive
    var indexTip: LandmarkPoint?
    var markers: [DetectedMarker] = []
    var sourceWidth: Int = 0
    var sourceHeight: Int = 0
    var notice: String?

    var stage: ProductionStage {
        if camera == .permissionRequired || camera == .permissionDenied { return .cameraPermission }
        if camera == .error { return .cameraError }
        switch connection.status {
        case .error: return .connectionError
        case .reconnecting: return .reconnecting
        case .connecting, .awaitingAck: return connection.automatic ? .autoConnecting : .connecting
        case .disconnected: return .connect
        case .connected:
            return captureMode == .calibration ? .calibration : .ready
        }
    }

    var visualState: ProductionVisualState {
        switch stage {
        case .cameraPermission: return .cameraPermission
        case .cameraError: return .cameraError
        case .connect, .connectionError: return .connectionForm
        case .connecting, .autoConnecting: return .connectionProgress
        case .reconnecting: return .reconnecting
        case .calibration:
            return calibration == .placementWaiting ? .placement : .calibrationProgress
        case .ready:
            switch tracking {
            case .candidate, .tracking, .temporarilyLost, .longLost: return .readyActive
            default: return .readyIdle
            }
        }
    }
}
